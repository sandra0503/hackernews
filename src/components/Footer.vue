<template>
  <footer class="footer">2019 Hackernews</footer>
</template>

<script>
export default {
  name: "Footer"
};
</script>

<style lang="scss" scoped="true">
.footer {
  background: #009688;
  color: #fff;
  font-size: 14px;
  padding: 20px;
}
</style>
