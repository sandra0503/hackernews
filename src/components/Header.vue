<template>
  <header class="header">Hackernews Top 500</header>
</template>

<script>
export default {
  name: "Header"
};
</script>

<style lang="scss" scoped="true">
.header {
  background: #009688;
  color: #fff;
  font-size: 24px;
  font-weight: 600;
  padding: 20px;
}
</style>
