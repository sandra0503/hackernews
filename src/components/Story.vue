<template>
  <div class="story" @click="onClick" :class="{ 'story--active': isActive }">
    <div class="story__score">
      <span class="story__upvote" @click.stop.prevent="onUpvote"/>
      <span v-if="score" class="story__score-value">{{ score }}</span>
    </div>
    <div class="story__content">
      <div v-if="strippedTitle" class="story__title">{{ strippedTitle }}</div>
      <div class="story__author">
        by
        <span v-if="author" class="story__author-name">{{ author }}</span>
      </div>
      <div v-if="strippedText" class="story__text">{{ strippedText }}</div>
      <a v-if="linkUrl" class="story__link" target="__blank" :href="linkUrl"
        >{{ linkUrl }}
    </a>
    </div>
  </div>
</template>

<script>
export default {
  name: "Story",
  data() {
    return {
      isActive: false
    };
  },
  props: {
    title: {
      type: String,
      required: true
    },
    author: {
      type: String
    },
    score: {
      type: Number
    },
    url: {
      type: String
    },
    text: {
      type: String
    }
  },
  computed: {
    strippedTitle() {
      if (!this.isActive && this.title.length > 100) {
         return `${this.title.slice(0, 99)}...`;
      }
      return this.title;
    },
    strippedText() {
      if (!this.isActive && this.text && this.text.length > 70) {
         return `${this.text.slice(0, 69)}...`;
      }
      return this.text;
    },
    linkUrl() {
      return this.isActive ? this.url : null;
    }
  },
  methods: {
    onClick() {
      this.isActive = !this.isActive;
    },
    onUpvote() {
      // TODO
    }
  }
};
</script>

<style lang="scss" scoped="true">
.story {
  display: flex;
  padding: 10px;
  border: 1px solid #dadce0;
  border-radius: 5px;
  cursor: pointer;
  overflow: hidden;
  transition: background .2s;

  &:hover {
    background: #f8f9fa;
  }

  &.story--active {
    grid-area: span 2 / span 2;
    background: #f8f9fa;
  }
}

.story__score {
  width: 26px;
  display: flex;
  flex-direction: column;
  text-align: center;
  margin-right: 10px;
  font-size: 12px;
}

.story__upvote {
  padding: 5px;
  margin-top: -15px;
  align-self: center;

  &:before {
    content: '';
    display: inline-flex;
    width: 0;
    height: 0; 
    border-left: 8px solid transparent;
    border-right: 8px solid transparent;
    border-bottom: 16px solid #009688;
    transition: border .2s;
  }

  &:hover {
    &:before {
      border-bottom-color: #00ccb8;
    }
  }
}

.story__content {
  font-size: 14px;
}

.story__title {
  font-size: 14px;
  font-weight: 500;
}

.story__link {
  word-break: break-all;
  text-decoration: none;
  color: #009688;

  &:hover {
    color: #00ccb8;
  }
}

.story__author {
  color: #808080;
  margin-bottom: 5px;
}

.story__text {
  margin-bottom: 10px;
}
</style>
