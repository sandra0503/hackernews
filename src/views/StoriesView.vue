<template>
  <div class="stories-view">
    <Header />
    <main class="stories">
      <Story
        v-for="story in stories"
        :key="story.id"
        :author="story.by"
        :score="story.score"
        :text="story.text"
        :title="story.title"
        :url="story.url"
      />
    </main>
    <Footer />
  </div>
</template>

<script>
import { mapState } from "vuex";
import Header from "../components/Header.vue";
import Footer from "../components/Footer.vue";
import Story from "../components/Story.vue";

export default {
  name: "StoriesView",
  components: {
    Header,
    Footer,
    Story
  },
  computed: {
    ...mapState({
      stories: "stories"
    })
  }
};
</script>

<style lang="scss">
.stories {
  padding: 20px;
  display: grid;
  grid-gap: 10px;
  grid-auto-flow: dense;
  grid-auto-rows: 120px;
}

@media only screen and (min-width: 800px) {
  .stories {
    grid-template-columns: repeat(4, 1fr);
    grid-auto-rows: 160px;
  }
}

@media only screen and (min-width: 1200px) {
  .stories {
    grid-template-columns: repeat(6, 1fr);
    grid-auto-rows: 160px;
  }
}
</style>
