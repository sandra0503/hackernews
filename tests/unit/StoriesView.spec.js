import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import StoriesView from "@/views/StoriesView";
import Story from "@/components/Story";
import Header from "@/components/Header";
import Footer from "@/components/Footer";
import initialState from "@/store/state";
import storyMock from "@/mocks/story";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("StoriesView", () => {
  let state;

  const build = () => {
    const wrapper = shallowMount(StoriesView, {
      localVue,
      store: new Vuex.Store({ state })
    });

    return {
      wrapper,
      story: () => wrapper.find(Story),
      header: () => wrapper.find(Header),
      footer: () => wrapper.find(Footer)
    };
  };

  beforeEach(() => {
    state = { ...initialState };
  });

  it("renders the component", () => {
    const { wrapper } = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders child components", () => {
    state.stories = [storyMock];
    const { story, header, footer } = build();

    expect(story().exists()).toBe(true);
    expect(header().exists()).toBe(true);
    expect(footer().exists()).toBe(true);
  });

  it("passes binded content props to content item components", () => {
    state.stories = [storyMock];
    const { story } = build();

    expect(story().vm.title).toBe(state.stories[0].title);
    expect(story().vm.text).toBe(state.stories[0].text);
    expect(story().vm.author).toBe(state.stories[0].by);
    expect(story().vm.url).toBe(state.stories[0].url);
    expect(story().vm.score).toBe(state.stories[0].score);
  });
});
