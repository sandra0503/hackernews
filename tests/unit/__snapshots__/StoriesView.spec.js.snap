// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoriesView renders the component 1`] = `
<div class="stories-view">
  <header-stub></header-stub>
  <main class="stories">
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html" text="lorem ipsum"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html" text="lorem ipsum"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html" text="lorem ipsum"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
    <story-stub title="My YC app: Dropbox - Throw away your USB drive" author="dhouston" score="104" url="http://www.getdropbox.com/u/2/screencast.html"></story-stub>
  </main>
  <footer-stub></footer-stub>
</div>
`;
